rag-toolchain/src/toolchain_indexing/types.rs

```diff
@@ -0,0 +1,118 @@
+use std::rc::Rc;
+
+// ----------------- Embedding -----------------
+pub struct Embedding {
+    embedding: Rc<[f32]>,
+}
+
+impl Embedding {
+    pub fn new(embedding: Rc<[f32]>) -> Self {
+        Self { embedding }
+    }
+
+    pub fn embedding(&self) -> &Rc<[f32]> {
+        &self.embedding
+    }
+
+    pub fn into_vec<T>(iter: impl Iterator<Item = T>) -> Vec<Self>
+    where
+        T: Into<Self>,
+    {
+        let mut vec: Vec<Embedding> = Vec::new();
+        for item in iter {
+            vec.push(item.into());
+        }
+        vec
+    }
+}
+
+impl From<Vec<f32>> for Embedding {
+    fn from(embedding: Vec<f32>) -> Self {
```

Comment:
These from implementations could probably just be generic with some trait bounds